websockets.md

```diff
@@ -0,0 +1,101 @@
+    CredentialCache.DefaultCredentials :
+    credentials;
+if (proxy == null)
+{
+    handler.UseProxy = false;
+}
+else
+{
+    handler.Proxy = proxy;
+}
+if (clientCertificates?.Count > 0)
+{
+    handler.ClientCertificates.AddRange(clientCertificates);
+}
+var invoker = new HttpMessageInvoker(handler);
+
+var ws = new ClientWebSocket();
+await ws.ConnectAsync(uri, invoker, cancellationToken);
+```
+
+## WebSockets over HTTP/2
+
+.NET 7 also adds an ability to use WebSocket protocol over HTTP/2, as described in [RFC 8441](https://www.rfc-editor.org/rfc/rfc8441). With that, WebSocket connection would be established over a single stream of HTTP/2 connection. This will allow for a single TCP connection to be shared between two protocols and to be used for several WebSocket connections and HTTP requests at the same time, resulting in more efficient use of the network.
+
+To enable WebSockets over HTTP/2, you can set [ClientWebSocketOptions.HttpVersion](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocketoptions.httpversion?view=net-7.0) option to `HttpVersion.Version20`. You can also enable upgrade/downgrade of HTTP version used by setting [ClientWebSocketOptions.HttpVersionPolicy](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocketoptions.httpversionpolicy?view=net-7.0) property. These options will behave in the same way `HttpRequestMessage.Version` and `HttpRequestMessage.VersionPolicy` behave.
+
+For example, the following code would probe for HTTP/2 WebSockets, and if a WebSocket connection could not be established, it will fallback to HTTP/1.1:
+
+```c#
+var ws = new ClientWebSocket();
+ws.Options.HttpVersion = HttpVersion.Version20;
+ws.Options.HttpVersionPolicy = HttpVersionPolicy.RequestVersionOrLower;
+await ws.ConnectAsync(uri, httpClient, cancellationToken);
+```
+
+The combination of `HttpVersion.Version11` and `HttpVersionPolicy.RequestVersionOrHigher` will result in the same behavior as above, while `HttpVersionPolicy.RequestVersionExact` will disallow upgrade/downgrade of HTTP version.
+
+By default, `HttpVersion = HttpVersion.Version11` and `HttpVersionPolicy = HttpVersionPolicy.RequestVersionOrLower`, which means that by default only HTTP/1.1 will be used.
+
+Note that, because the ability to multiplex WebSocket connections (and HTTP requests) over a single HTTP/2 connection is the crucial part of the feature, you can *only* use it when you pass an `HttpClient` of `HttpMessageInvoker` instance from your code (i.e. use [ClientWebSocket.ConnectAsync(Uri, HttpMessageInvoker, CancellationToken)](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocket.connectasync?view=net-7.0#system-net-websockets-clientwebsocket-connectasync(system-uri-system-net-http-httpmessageinvoker-system-threading-cancellationtoken)) overload), so that the connection pool within this `HttpClient` of `HttpMessageInvoker` instance could be used for multiplexing.
```

Comment:
```suggestion
The ability to multiplex WebSocket connections and HTTP requests over a single HTTP/2 connection is a crucial part of this feature. For it to work as expected, you need to pass and reuse the same `HttpMessageInvoker` instance (e.g. `HttpClient`) from your code when calling `ConnectAsync`, i.e. use [ClientWebSocket.ConnectAsync(Uri, HttpMessageInvoker, CancellationToken)](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.clientwebsocket.connectasync?view=net-7.0#system-net-websockets-clientwebsocket-connectasync(system-uri-system-net-http-httpmessageinvoker-system-threading-cancellationtoken)) overload. This will reuse the connection pool within the `HttpMessageInvoker` instance for the multiplexing.
```

Not sure this is much better, but "`HttpClient` of `HttpMessageInvoker` instance" is a cumbersome phrase. Also this might be better accompanied with an example how to set and reuse the client.